gcp-common/src/main/java/io/micronaut/gcp/credentials/AuthenticationLoggingInterceptor.java

```diff
@@ -0,0 +1,93 @@
+import io.micronaut.core.type.MutableArgumentValue;
+import jakarta.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An interceptor for managed instances of {@link com.google.auth.oauth2.GoogleCredentials} that logs certain types of
+ * authentication errors that the GCP libraries handle silently as infinitely retryable events.
+ *
+ * @author Jeremy Grelle
+ * @since 5.2.0
+ */
+@Singleton
+public class AuthenticationLoggingInterceptor implements MethodInterceptor<Object, Object> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AuthenticationLoggingInterceptor.class);
+    private static final String LOGGED_AUTHENTICATION_METHOD = "getRequestMetadata";
+
+    /**
+     * Intercepts the "getRequestMetadata" call and logs any retryable errors before allowing the GCP library to continue
+     * its normal retry algorithm.
+     *
+     * @param context The method invocation context
+     * @return the result of the method invocation
+     */
+    @Override
+    public @Nullable Object intercept(MethodInvocationContext<Object, Object> context) {
+        if (!context.getExecutableMethod().getMethodName().equals(LOGGED_AUTHENTICATION_METHOD)) {
+            return context.proceed();
+        }
+        Map<String, MutableArgumentValue<?>> params = context.getParameters();
+        params.entrySet().stream().filter(entry -> entry.getValue().getType().equals(RequestMetadataCallback.class))
+            .findFirst()
+            .ifPresent(entry -> {
+                @SuppressWarnings("unchecked") MutableArgumentValue<RequestMetadataCallback> argValue = (MutableArgumentValue<RequestMetadataCallback>) entry.getValue();
+                RequestMetadataCallback callback = argValue.getValue();
+                argValue.setValue(new LoggingRequestMetadataCallback(callback));
+            });
+        return context.proceed();
+    }
+
+    /**
+     * A wrapper {@link RequestMetadataCallback} implementation that logs failures with a warning before proceeding with
+     * the original callback.
+     */
+    private static final class LoggingRequestMetadataCallback implements RequestMetadataCallback {
+
+        private final RequestMetadataCallback callback;
+
+        private LoggingRequestMetadataCallback(RequestMetadataCallback callback) {
+            this.callback = callback;
+        }
+
+        @Override
+        public void onSuccess(Map<String, List<String>> metadata) {
+            this.callback.onSuccess(metadata);
+        }
+
+        @Override
+        public void onFailure(Throwable ex) {
+            if (ex instanceof IOException) {
+                LOG.warn("A failure occurred while attempting to build credential metadata for a GCP API request. The GCP libraries treat this as " +
+                    "a retryable error, but misconfigured credentials can keep it from ever succeeding.", ex);
```

Comment:
I wonder if the logging should be DEBUG instead of WARN. If Google do it silently, maybe it's a common occurrence, and WARN might end up spamming people's dashboards, without them being able to easily turn it down (without losing any info logs they have) 🤔

Reply:
The GCP library logs it internally at the equivalent of DEBUG level - but turning that on results in a ton of extraneous noise and is a pain to do in the first place because of their use of JUL.

I tested a lot of different scenarios when digging into the source issue (#687) and determined that a failure at this particular point is almost always because the credentials are actually misconfigured and thus this call will never succeed until the credentials are corrected - which is possible to do in GCP's IAM management console at runtime, for example if the given service account is just missing the sufficient permission level to subscribe to the PubSub topic. I believe that's the reasoning as to why they treat it as a retry-able failure.

Not logging this occurrence by default can lead to a lot of tail-chasing in trying to figure out why your subscribers aren't receiving any published messages, so I think it's better that we explicitly show the failure. Note that the source exceptions (which we include in the log) are actually quite useful in showing exactly what went wrong. If someone finds it too noisy, they can always turn down the logging level for this class. ;)

Reply:
Good call 👍 

I always worry about error logs as they are the ones that get people out of bed 😉